keps/sig-aws/0014-iam-credentials-provider.md

```diff
@@ -0,0 +1,209 @@
+
+Users are able to associate the [IAM Role](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles.html) that the Pod will be able to assume.
+
+### Implementation Details/Notes/Constraints
+
+1. Solution should work for existing applications that use an AWS SDK without requiring application code changes 
+2. Cluster operators can run secure clusters by minimising IAM permissions needed on each node
+3. Pods that need IAM credentials shouldn’t start until credentials are available
+4. IAM messages/errors are accessible to cluster users
+5. Credentials should be auditable against strongest Pod identity that can be verified
+6. Credentials should be time-limited and refresh, allowing compromised credentials to be quickly invalidated
+
+Our suggested implementation is composed of:
+
+#### IAMBinding Custom Resource.
+This will contain a spec associating a Service Account with an AWS IAM Role.
+
+For example:
+
+```yaml
+---
+apiVersion: v1
+kind: Pod
+spec:
+  serviceAccountName: reporting-app
+  ...
+
+---
+apiVersion: iam.aws.k8s.io/v1alpha1
+kind: IAMBinding
+metadata:
+  name: reporting-role
+spec:
+  aws:
+    roleArn: arn:aws:iam::123456789012:role/S3Access
+  serviceAccountRef:
+    name: reporting-app
+```
+
+#### Server
+This process will verify the identity of the Pod and determine which IAM credentials it’s permitted to obtain, issue them using Amazon STS and return them. 
+
+By setting a role session name to <namespace>.<service-account-name> (potentially with a cluster-specific suffix provided via a command-line flag** when obtaining credentials, we can use it in an assume policy to restrict which namespace or service account can use the role.
+
+Pods will be associated to IAM roles using a Custom Resource: an IAMBinding. This will map between a Pod’s service account and the IAM role.
+
+**Restricting roles to namespaces**
+
+Our proposal would allow IAM assume policy to be set similarly to below for controlling which service accounts can assume which roles:
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Sid": "",
+      "Action": "sts:AssumeRole",
+      "Effect": "Allow",
+      "Principal": {
+        "AWS": "arn:aws:iam::123456789012:role/kiam_role"
+      },
+      "Condition": {
+        "StringLike": {
+          "aws:userid": "cluster.namespace.service-account-name"
+        }
+      }
+    }
+  ]
+}
+```
+
+#### Proxy
+Any Pod needing credentials will have an init container and sidecar. 
+
+The init container will ensure that other containers in the Pod when accessing `http://169.254.170.2` will have requests forwarded to the sidecar. The init container can also request initial credentials ensuring that the pod is entitled to use credentials for a role configured through the IAMBinding.
```

Comment:
> The init container will ensure that other containers in the Pod when accessing `http://169.254.170.2` will have requests forwarded to the sidecar.

can you describe the mechanism used for that? what level of permission/privilege is required for the init container to accomplish this?

Reply:
Within Istio I believe it's done with iptables here: https://github.com/istio/istio/blob/0759c8572119b8c78eeed1e0bbd3c58efc748092/tools/deb/istio-iptables.sh

We do similar with an agent proxy already which adds `NET_ADMIN` to the security context capabilities. So for this proposal it _could_ be done using the same.